src/server/terrain/structure_building_blocks.zig

```diff
@@ -0,0 +1,288 @@
+const std = @import("std");
+
+const main = @import("root");
+const ZonElement = main.ZonElement;
+const Blueprint = main.blueprint.Blueprint;
+const List = main.List;
+const AliasTable = main.utils.AliasTable;
+const Neighbor = main.chunk.Neighbor;
+const Block = main.blocks.Block;
+const parseBlock = main.blocks.parseBlock;
+const Degrees = main.rotation.Degrees;
+const hashInt = main.utils.hashInt;
+const NeverFailingAllocator = main.heap.NeverFailingAllocator;
+
+var arena = main.heap.NeverFailingArenaAllocator.init(main.globalAllocator);
+const arenaAllocator = arena.allocator();
+
+var structureCache: ?std.StringHashMapUnmanaged(StructureBuildingBlock) = null;
+var blueprintCache: ?std.StringHashMapUnmanaged([4]*BlueprintEntry) = null;
+var childrenToResolve: List(struct {parentId: []const u8, colorName: []const u8, colorIndex: usize, childIndex: usize, structureId: []const u8}) = undefined;
+
+const BlueprintEntry = struct {
+	blueprint: Blueprint,
+	info: Info,
+
+	fn init(blueprint: Blueprint, stringId: []const u8) !*BlueprintEntry {
+		const self = arenaAllocator.create(BlueprintEntry);
+		self.* = .{
+			.blueprint = blueprint,
+			.info = try Info.initFromBlueprint(arenaAllocator, blueprint, stringId),
+		};
+		return self;
+	}
+
+	const Info = struct {
+		const StructureBlock = struct {
+			x: i32,
+			y: i32,
+			z: i32,
+			index: u32,
+			block: Block,
+
+			pub inline fn direction(self: StructureBlock) Neighbor {
+				return @enumFromInt(self.block.data);
+			}
+		};
+
+		originBlock: StructureBlock,
+		childrenBlocks: List(StructureBlock),
```

Comment:
Should be ListUnmanaged in anticipation of #1181

Reply:
Actually this should be a slice, since after loading the number of elements is known.

Reply:
This is a list of blocks that appear in blueprint. Single color child block may appear multiple times. Only way to determine this value is to do one detection pass through whole blueprint and count how many child blocks there are. Will be as requested.